ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java

```diff
@@ -114,7 +114,85 @@ public void start() throws StoreException {
+      for (BlobId blobId : blobIdList) {
+        CloudBlobMetadata blobMetadata = cloudBlobMetadataListMap.get(blobId.getID());
+        MessageInfo messageInfo = new MessageInfo(blobId, blobMetadata.getSize(), blobMetadata.getExpirationTime(),
+            (short) blobMetadata.getAccountId(), (short) blobMetadata.getContainerId(), getOperationTime(blobMetadata));
+        messageInfos.add(messageInfo);
+        blobReadInfos.add(new CloudMessageReadSet.BlobReadInfo(blobMetadata, blobId));
+      }
+    } catch (CloudStorageException e) {
+      throw new StoreException(e, StoreErrorCodes.IOError);
+    }
+    CloudMessageReadSet messageReadSet = new CloudMessageReadSet(blobReadInfos, cloudDestination);
+    return new StoreInfo(messageReadSet, messageInfos);
+  }
+
+  /**
+   * validate the {@code CloudBlobMetadata} map to make sure it has metadata for all keys, and they meet the {@code storeGetOptions} requirements.
+   * @param cloudBlobMetadataListMap
+   * @throws StoreException if the {@code CloudBlobMetadata} isnt valid
+   */
+  private void validateCloudMetadata(Map<String, CloudBlobMetadata> cloudBlobMetadataListMap,
+      EnumSet<StoreGetOptions> storeGetOptions) throws StoreException {
+    for (String key : cloudBlobMetadataListMap.keySet()) {
+      if (isBlobDeleted(cloudBlobMetadataListMap.get(key)) && !storeGetOptions.contains(
+          StoreGetOptions.Store_Include_Deleted)) {
+        throw new StoreException("Id " + key + " has been deleted on the cloud", StoreErrorCodes.ID_Deleted);
+      }
+      if (isBlobExpired(cloudBlobMetadataListMap.get(key)) && !storeGetOptions.contains(
+          StoreGetOptions.Store_Include_Expired)) {
+        throw new StoreException("Id " + key + " has expired on the cloud", StoreErrorCodes.TTL_Expired);
+      }
+    }
+  }
+
+  /**
+   * Gets the operation time for a blob from blob metadata based on the blob's current state and timestamp recorded for that state.
+   * @param metadata blob metadata from which to derive operation time.
+   * @return operation time.
+   */
+  private long getOperationTime(CloudBlobMetadata metadata) {
+    if (isBlobDeleted(metadata)) {
+      return metadata.getDeletionTime();
+    } else if (isBlobExpired(metadata)) {
+      return metadata.getExpirationTime();
+    }
+    return metadata.getCreationTime();
+  }
+
+  /**
+   * Check if the blob is marked for deletion in its metadata
+   * @param metadata to check for deletion
+   * @return true if deleted. false otherwise
+   */
+  static boolean isBlobDeleted(CloudBlobMetadata metadata) {
+    return metadata.getDeletionTime() != Utils.Infinite_Time;
+  }
+
+  /**
+   * Check if the blob is expired
+   * @param metadata to check for expiration
+   * @return true if expired. false otherwise
+   */
+  static boolean isBlobExpired(CloudBlobMetadata metadata) {
+    return metadata.getExpirationTime() != Utils.Infinite_Time
+        && metadata.getExpirationTime() < System.currentTimeMillis();
```

Comment:
Not good to get system time in method that is called in a loop.  Please pass system time to method and call once at top of loop.

Reply:
fixed.